theseus/optimizer/nonlinear/dcem.py

```diff
@@ -0,0 +1,247 @@
+            vectorize=vectorize,
+            abs_err_tolerance=abs_err_tolerance,
+            rel_err_tolerance=rel_err_tolerance,
+            max_iterations=max_iterations,
+            **kwargs,
+        )
+
+        self.objective = objective
+        self.ordering = VariableOrdering(objective)
+        self.n_samples = n_sample
+        self.n_elite = n_elite
+        self.lb = lb
+        self.ub = ub
+        self.temp = temp
+        self.normalize = normalize
+        self._tot_dof = sum([x.dof() for x in self.ordering])
+        self.lml_eps = lml_eps
+        self.lml_verbose = lml_verbose
+        self.init_sigma = init_sigma
+
+    def _mu_vec_to_dict(self, mu: torch.Tensor) -> Dict[str, torch.Tensor]:
+        idx = 0
+        mu_dic = {}
+        for var in self.ordering:
+            mu_dic[var.name] = mu[:, slice(idx, idx + var.dof())]
+            idx += var.dof()
+        return mu_dic
+
+    def reset_sigma(self, init_sigma: Union[float, torch.Tensor]) -> None:
+        self.sigma = (
+            torch.ones(
+                (self.objective.batch_size, self._tot_dof), device=self.objective.device
+            )
+            * init_sigma
+        )
+
+    def _CEM_step(self):
+        """
+        Performs one iteration of CEM.
+        Updates the self.sigma and return the new mu.
+        """
+        device = self.objective.device
+        n_batch = self.ordering[0].shape[0]
+
+        mu = torch.cat([var.tensor for var in self.ordering], dim=-1)
+
+        X = Normal(mu, self.sigma).rsample((self.n_samples,))
+
+        X_samples: List[Dict[str, torch.Tensor]] = []
+        for sample in X:
+            X_samples.append(self._mu_vec_to_dict(sample))
+
+        fX = torch.stack(
+            [self.objective.error_metric(X_samples[i]) for i in range(self.n_samples)],
+            dim=1,
+        )
+
+        assert fX.shape == (n_batch, self.n_samples)
+
+        if self.temp is not None and self.temp < np.infty:
+            if self.normalize:
+                fX_mu = fX.mean(dim=1).unsqueeze(1)
+                fX_sigma = fX.std(dim=1).unsqueeze(1)
+                _fX = (fX - fX_mu) / (fX_sigma + 1e-6)
+            else:
+                _fX = fX
+
+            if self.n_elite == 1:
+                # indexes = LML(N=n_elite, verbose=lml_verbose, eps=lml_eps)(-_fX*temp)
+                indexes = torch.softmax(-_fX * self.temp, dim=1)
+            else:
+                indexes = LML(
+                    N=self.n_elite, verbose=self.lml_verbose, eps=self.lml_eps
+                )(-_fX * self.temp)
+            indexes = indexes.unsqueeze(2)
+            eps = 0
+
+        else:
+            indexes_vals = fX.argsort(dim=1)[:, : self.n_elite]
+            # Scatter 1.0 to the indexes using indexes_vals
+            indexes = torch.zeros(n_batch, self.n_samples, device=device).scatter_(
+                1, indexes_vals, 1.0
+            )
+            # indexes = torch.zeros(n_batch, self.n_samples, device=device)
```

Comment:
Can remove the code that is commented out.

Reply:
done